crafting.py

```diff
@@ -0,0 +1,316 @@
+   
+        
+    return propositions 
+
+    
+def preconditions_satisfied(state: State, recipe: Recipe) -> bool:
+  
+    if ( state >= recipe.requires and state >= recipe.consumes) :
+        return True
+    
+    return False
+
+
+
+def apply_effects(state: State, recipe: Recipe) -> State:
+    # E. How do you make a new state out of a state and a recipe?
+    # Note, DO NOT change state in-place!
+     newstate  =  state - recipe.consumes + recipe.produces
+    
+     return newstate 
+ 
+def see_state(state:State, combinations:List[Set[Proposition]], seen_combinations:Set[FrozenSet[Proposition]]) -> bool:
+    any_new = False
+    state_props = state_propositions(state)
+    for combo in combinations:
+        if combo in seen_combinations:
+            continue 
+        
+        if(state_props.issuperset(combo)) :
+            seen_combinations.add(combo)
+            any_new = True
+        
+    return any_new
+
+   
+recipe_propositions = set()
+for r in recipes.values():
+    recipe_propositions |= recipe_to_propositions(r)    
+
+def plan_dijkstra(initial: State, goal: State, limit:int) -> Tuple[int, int, Optional[List[str]]]:
+    start_time = time.time()
+    open_list: List[Tuple[int, State,str]] = [(0, initial, "")]         
+    best_costs: Dict[State, Tuple[int, str,State]] = {initial:(0, "",initial)}
+    visit_count = 0
+    found = False
+    while open_list:
+        # Dijkstra's search uses the priority queue data structure
+        cost, curstate , gotfrom  = heapq.heappop(open_list)       
+        #print("cost=",cost,"curstate=",curstate,"from=",gotfrom)
+       
+        if(curstate >= goal):
+            x,rname,z = best_costs.get(curstate)
+            print ("found it!! Number of visited nodes=" ,visit_count , "state=",curstate,"goal=",goal,"by=",rname) 
+            found = True;
+            break;
+                
+        visit_count += 1
+        if(visit_count > limit) :
+            print("too many iterations , current state=",curstate)            
+            break
+        
+                         
+        for candidaterecipename in recipes:
+            if (preconditions_satisfied(curstate,recipes[candidaterecipename])) == False:
+                continue 
+            
+            candidatestate = apply_effects(curstate,recipes[candidaterecipename])
+            
+            if(candidatestate in  best_costs):
+                #print("found=" , neighbor)     
+                candidatestate_cost,rname,fromstate = best_costs.get(candidatestate)
+                if(candidatestate_cost > cost + recipes[candidaterecipename].cost) :        
+                    #print("update cost for ", neighbor ,"newcost", cost + neighbor_cost)
+                    best_costs[candidatestate] = (cost + recipes[candidaterecipename].cost ,  gotfrom, curstate )   
+                    heapq.heappush (open_list,  (cost + recipes[candidaterecipename].cost, candidatestate,  candidaterecipename ) )     
```

Comment:
Using `dijkstra`'s algorithm, don't we not need to push to the `open_list` if the item already in `best_costs`?

Reply:
What happen if  you have a  better cost now  to get to the node ? My thinking you want to reconsider this  path again given the new cost

Reply:
I thought how Dijkstra's worked is that since it is optimal, as soon as a node is removed from the open_list and visited that we have found the best path to that node, like when we find the goal we don't have to keep searching for another possible better path. So with this understanding, there is no need to add it back to open_list since the path we have found is guaranteed optimal. 

This is not the case for BFS, which is why we add it back in that scenario.

Reply:
The node which we are comparing against best_costs is not the one we just visited on the open list.  Consider this graph with start state s and goal state g:
```
s-(100)->g
s-(1)->a
a-(1)->g
```
You’ll visit s and get paths to a and g, then visit a and have to update your path to g.
It’s easy to imagine a variation on this graph where g is not the goal, but is one of two nodes which can get to the goal, and is the only one of those two along an optimal path.  In that case not adding g to the queue again would prevent finding the optimal path.

It’s true that once you visit a node you know your path there was optimal, so maybe the terminology of reopening is not helpful; more like, add it to the open set again at a lower priority.

Does that make sense?

Reply:
I see thanks!  So the code is correct after all ?

Reply:
I think I see what is happening. Yes, the code is correct - it is trying to update the priority of that element on the heap, just like you are updating the `best_cost` when you have found a better path. I have a question then, why use a heap, where we can't update priority values and have to now have two elements (or more) to later remove? 

We are all 👍!

Reply:
The nice thing about Dijkstra’s algorithm is that, as you say, once you open a node you know you have the best path to it.  You might try as an optimization ignoring a node you’ve already opened, either with a closed set or by comparing the node priority vs the best known path.  But not doing that won’t lead to any cycles because you won’t add any of the newly re-discovered neighbors to the open set.
Removing an item from some unknown spot in the priority queue is very expensive compared to removing just the front item, because you have to find the item first! And much of the time, the item won’t be in there anyway...